CubeSQL_SDK/cubesql.c

```diff
@@ -936,185 +937,238 @@ void csql_dbfree (csqldb *db) {
+		ioctl(sock_current, FIONBIO, &ioctl_blocking);
+		
+		// initiate non-blocking connect ignoring return code
+		rc = connect(sock_current, addr->ai_addr, addr->ai_addrlen);
+		if (rc < 0) {
+			if (errno != 0 && errno != EINTR && errno != EAGAIN & errno != EINPROGRESS)
+			{
+				// obvious error (e.g. network not reachable) - don't use this socket
+				lastConnectionErrorMessage = strerror(errno);
+				/*
+				printf("connect error: %s (%i)\n", strerror(errno), errno);
+				*/
+				closesocket(sock_current);
+				continue;
+			}
+		}
+		
+		// add sock_current to internal list of trying to connect sockets
+		sock_list[sock_index] = sock_current;
+	}
+	
+	// free not more needed memory
+	freeaddrinfo(addr_list);
 	
 	// calculate the connection timeout and reset timers
 	int connect_timeout = (db->timeout > 0) ? db->timeout : CUBESQL_DEFAULT_TIMEOUT;
 	time_t start = time(NULL);
-    time_t now = start;
-    rc = 0;
-    
-    int socket_err = 0;
-    int sockfd = 0;
-    fd_set write_fds;
-    fd_set except_fds;
-    struct timeval tv;
-    
-	while (rc == 0 && ((now - start) < connect_timeout)) {
+	time_t now = start;
+	rc = 0;
+	
+	int sockfd = 0;
+	fd_set write_fds;
+	fd_set except_fds;
+	struct timeval tv;
+
+	while ((now - start) < connect_timeout) {
 		FD_ZERO(&write_fds);
-        FD_ZERO(&except_fds);
-        
-        int nfds = 0;
-        for (int i=0; i<MAX_SOCK_LIST; ++i) {
-            if (sock_list[i]) {
-                FD_SET(sock_list[i], &write_fds);
-                FD_SET(sock_list[i], &except_fds);
-                if (nfds < sock_list[i]) nfds = sock_list[i];
-            }
-        }
-        
+		FD_ZERO(&except_fds);
+		
 		tv.tv_sec = connect_timeout;
 		tv.tv_usec = 0;
+
+		for (int i=0; i<MAX_SOCK_LIST; ++i) {
+			if (sock_list[i] > 0) {
+				FD_SET(sock_list[i], &write_fds);
+				FD_SET(sock_list[i], &except_fds);
+
+				rc = bsd_select(sock_list[i] + 1, NULL, &write_fds, &except_fds, &tv);
+				
+				if (rc == 0) {
+					// timeout - we can't use this socket
+					closesocket(sock_list[i]);
+					sock_list[i] = -1;
+					continue;
+				}
+
+				if (rc < 0) {
+					if (errno == 0 || errno == EINTR || errno == EAGAIN || errno == EINPROGRESS) {
+						// socket is still connecting
+						continue;
+					}
+					
+					// socket bsd_select error - don't use this socket
+					/*
+					printf("socket bsd_select error: %s (%i)\n", strerror(errno), errno);
+					*/
+					lastConnectionErrorMessage = strerror(errno);
+					closesocket(sock_list[i]);
+					sock_list[i] = -1;
+					continue;
+				}
+			}
+		}
+		
+		// check for error
+		for (int i=0; i<MAX_SOCK_LIST; ++i) {
+			if (sock_list[i] > 0) {
+				if (FD_ISSET(sock_list[i], &except_fds)) {
+					int err = csql_socketerror(sock_list[i]);
+					if (err > 0) {
+						lastConnectionErrorMessage = strerror(err);
+					}
+
+					closesocket(sock_list[i]);
+					sock_list[i] = -1;
+				}
+			}
+		}
 		
-		rc = bsd_select(nfds + 1, NULL, &write_fds, &except_fds, &tv);
```

Comment:
so only the 1 `nfds` (out of other possible sockets) is being checked for errors here (the others are ignored)